x-pack/libbeat/management/generate.go

```diff
@@ -0,0 +1,253 @@
+		if err != nil {
+			return nil, fmt.Errorf("Error injecting stream processors: %w", err)
+		}
+		streamSource, err = injectAgentInfoRule(streamSource, agentInfo)
+		if err != nil {
+			return nil, fmt.Errorf("Error injecting agent processors: %w", err)
+		}
+		inputs[iter] = streamSource
+	}
+
+	return inputs, nil
+}
+
+// CreateReloadConfigFromInputs turns a raw input/module list into the ConfigWithMeta type used by the reloader interface
+func CreateReloadConfigFromInputs(raw []map[string]interface{}) ([]*reload.ConfigWithMeta, error) {
+	// format for the reloadable list needed bythe cm.Reload() method
+	configList := make([]*reload.ConfigWithMeta, len(raw))
+
+	for iter := range raw {
+		uconfig, err := conf.NewConfigFrom(raw[iter])
+		if err != nil {
+			return nil, fmt.Errorf("error in conversion to conf.C: %w", err)
+		}
+		configList[iter] = &reload.ConfigWithMeta{Config: uconfig}
+	}
+	return configList, nil
+}
+
+// Emulates the InjectAgentInfoRule and InjectHeadersRule ast rules
+func injectAgentInfoRule(inputs map[string]interface{}, agentInfo *client.AgentInfo) (map[string]interface{}, error) {
+	// upstream API can sometimes return a nil agent info
+	if agentInfo == nil {
+		return inputs, nil
+	}
+	var processors = []interface{}{}
+
+	processors = append(processors, generateAddFieldsProcessor(
+		mapstr.M{"id": agentInfo.ID, "snapshot": agentInfo.Snapshot, "version": agentInfo.Version},
+		"elastic_agent"))
+	processors = append(processors, generateAddFieldsProcessor(
+		mapstr.M{"id": agentInfo.ID},
+		"agent"))
+
+	currentProcs, ok := inputs["processors"]
+	if !ok {
+		inputs["processors"] = processors
+	} else {
+		currentProcsList, ok := currentProcs.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf("error creating list of existing processors, got: %#v", currentProcs)
+		}
+		inputs["processors"] = append(processors, currentProcsList...)
+
+	}
+
+	return inputs, nil
+}
+
+// injectIndexStream is an emulation of the InjectIndexProcessor AST code
+// TODO: not sure the logic for grabbing fields from the input-level vs. stream-level DataStream fields is correct?
+func injectIndexStream(expected *proto.UnitExpectedConfig, inputType string, streamIter int, stream map[string]interface{}) map[string]interface{} {
+	streamType := expected.DataStream.Type
+	if streamType == "" {
+		streamType = inputType
+	}
+	dataset := expected.Streams[streamIter].DataStream.Dataset
+	if dataset == "" {
+		dataset = "generic"
+	}
+	namespace := expected.DataStream.Namespace
+	if namespace == "" {
+		namespace = "default"
+	}
+	index := fmt.Sprintf("%s-%s-%s", streamType, dataset, namespace)
+	stream["index"] = index
+	return stream
+}
+
+//injectStreamProcessors is an emulation of the InjectStreamProcessorRule AST code
+func injectStreamProcessors(expected *proto.UnitExpectedConfig, inputType string, streamIter int, stream map[string]interface{}) (map[string]interface{}, error) {
```

Comment:
There is a processor explicitly for doing this you can probably reuse: https://github.com/elastic/beats/blob/main/libbeat/processors/add_data_stream/add_data_stream.go